src/internal/m365/collection/drive/delta_tree.go

```diff
@@ -459,6 +461,111 @@ func walkTreeAndBuildCollections(
 	return nil
 }
 
+type idPrevPathTup struct {
+	id       string
+	prevPath string
+}
+
+// fuses the collectables and old prevPaths into a
+// new prevPaths map.
+func (face *folderyMcFolderFace) generateNewPreviousPaths(
+	collectables map[string]collectable,
+	prevPaths map[string]string,
+) (map[string]string, error) {
+	var (
+		// id -> currentPath
+		results = map[string]string{}
+		// prevPath -> currentPath
+		movedPaths = map[string]string{}
+		// prevPath -> {}
+		tombstoned = map[string]struct{}{}
+	)
+
+	// first, move all collectables into the new maps
+
+	for id, cbl := range collectables {
+		if cbl.currPath == nil {
+			tombstoned[cbl.prevPath.String()] = struct{}{}
+			continue
+		}
+
+		cp := cbl.currPath.String()
+		results[id] = cp
+
+		if cbl.prevPath != nil && cbl.prevPath.String() != cp {
+			movedPaths[cbl.prevPath.String()] = cp
+		}
+	}
+
+	// next, create a slice of tuples representing any
+	// old prevPath entry whose ID isn't already bound to
+	// a collectable.
+
+	unseenPrevPaths := []idPrevPathTup{}
+
+	for id, p := range prevPaths {
+		if _, ok := results[id]; !ok {
```

Comment:
nit: I think it would be more clear here if paths that are known to be tombstoned are also removed from `unseenPrevPaths` here instead of below. Technically those paths are bound to a collectable like the comment above mentions and they shouldn't be updated or added to the set of output paths

Reply:
hierarchy makes that a little cumbersome.  I'd have to walk up the directory in every previous entry looking for a parent that was tombstoned.  The way I have it now, the tombstone map grows in sync with the root-to-leaf ordering that we get from this list.  I think the result would be about on par in terms of clarity.

Reply:
I'm not sure I follow, and maybe I was a little unclear. I was talking about moving lines 524-527 below up to here. I don't believe that should require hierarchy traversal since it's a direct comparison on the path

Reply:
oh yeah, that works.  My brain is going off into the weeds.